csp_billing_adapter_microsoft/plugin.py

```diff
@@ -81,55 +79,75 @@ def get_account_info(config: Config):
     The information contains the metadata for compute and network.
     """
     account_info = _get_metadata()
-    account_info['compute'] = json.loads(account_info.get('compute', '{}'))
-    account_info['network'] = json.loads(account_info.get('network', '{}'))
     account_info['cloud_provider'] = get_csp_name(config)
 
     return account_info
 
 
-def get_api_header():
-    """
-    Get the header to be used in requests
+def _get_metadata():
+    """Return a dict containing compute, network and signature information."""
+    metadata = {}
+    try:
+        metadata = _get_instance_metadata()
+        metadata['attestedData'] = _get_signature()
+    except ValueError as error:
+        log.error('Could not load JSON from metadata: {}'.format(error))
+        for key in ['compute', 'network', 'signature', 'pkcs7']:
+            if not metadata.get(key):
+                metadata[key] = {}
+
+    return metadata
+
+
+def _get_instance_metadata():
+    """Return all compute and network information from metadata."""
+    api_latest_version = _get_latest_api_version()
+    return json.loads(
+        _fetch_metadata(
+            INSTANCE_INFO_URL.format(latest=api_latest_version)
+        )
+    )
 
-    Prefer IMDS which requires a token.
-    """
-    request = urllib.request.Request(
-        METADATA_TOKEN_URL,
-        headers={'Metadata': 'True'},
-        method='GET'
+
+def _get_signature():
+    """Return attested data signature from metadata."""
+    api_latest_version = _get_latest_api_version()
+    return json.loads(
+        _fetch_metadata(
+            SIGNATURE_URL.format(latest=api_latest_version)
+        )
     )
 
-    try:
-        token = urllib.request.urlopen(request).read().decode()
-    except urllib.error.URLError as error:
-        log.error(f'Failed to retrieve metadata token: {str(error)}')
-        return {}
 
-    return token
+def _get_latest_api_version():
+    """
+    Return the newest API version available
 
+    Otherwise, return minimum API version to get
+    license type on the signature.
+    """
+    versions = json.loads(_fetch_metadata(VERSIONS_URL))
+    if versions:
+        return versions['apiVersions'][-1]
 
-def _get_metadata():
-    metadata_options = ['compute', 'network']
-    return {
-        metadata_option: _fetch_metadata(metadata_option)
-        for metadata_option in metadata_options
-    }
+    # minimum API version with license type
+    # for more info check https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux#attested-data
+    return '2020-09-01'
```

Comment:
This implementation makes no sense. 
- There's no check that `versions['apiVersions'][-1]` >= '2020-09-01'. 
- If we cannot access `VERSIONS_URL` then how do we know what is actually available?
- If we know that '2020-09-01' is deployed everywhere then why even query for the latest version and not just use the version we know provides the information we need?
- Returning a static '2020-09-01' in case `VERSIONS_URL` is not available contradicts the name of the function as we do not actually know what the latest version is.